### DISCUSSION TITLE
7.5.7 does not actually fix the issue reported in #3969

### DISCUSSION THREAD

**Discussion by schneider-felix** [deployphp/deployer#3972] [Bugs] [answered]
### Deployer Version

v7.5.7

### Target OS

Docker PHP Image

### Which PHP version are you using?

PHP 8.2

### Steps to reproduce

The issue reported in #3969 is not actually solved in v7.5.7. 
This is because the underlying reason for the issues we are currently seeing is a different one than the order of parameters. With the new release, the issue now just shifted to the `Constraint.php` file which has the same Problem of optional parameters before required ones.

The root cause for all these issues can actually be found elsewhere entirely. The Exceptions thrown by these files are actually just PHP Deprecations. Nothing that should break the tool entirely. However in https://github.com/deployphp/deployer/blob/dbc39848a458ac7957d1975cb98808ea2ee87268/bin/dep#L86-L93 these Deprecations are promoted to ErrorExceptions when the error_reporting value is configured accordingly. 

I think this should be changed to just write to STDERR for anything that's not at least an Error. Otherwise deployer will just always break when new PHP releases or dependencies cause deprecations. 

Lastly, the error is not caused by the dependencies themselves but instead because PHPCSFixer runs on the vendor directory. I don't really think dependency code should ever be touched by tools like CSFixer but that's just my opinion. 

Hope this somehow helped.

**Comment by antonmedv** [comment, marked as answer]
True. I will add skipping deprications:

```
    if ($severity == E_DEPRECATED) {
        return;
    }
```
And I fixed incrementPath. I did check remaining files with:

```
grep -irnE --include \*.php 'function.+?=.+?\$[^=]+\)'
grep -irnE --include \*.php 'function.+?=.+?\$[^=]+\,'
```

**Reply by schneider-felix** [reply]
Nice, thank you. Don't forget about `E_USER_DEPRECATED`. Maybe even consider doing the same for warnings and notices to prevent hard crashes in those cases. Maybe something like this could work to inform the user about possible issues while still ensuring the code doesn't just hard crash:

```php
if ($severity === E_DEPRECATED || $severity === E_USER_DEPRECATED || $severity === E_WARNING || $severity === E_USER_WARNING || $severity === E_NOTICE || $severity === E_USER_NOTICE) {
    fwrite(STDERR, $message);
    return;
}
```

**Reply by antonmedv** [reply]
Good point. But I already did the release)
